fix(core): dynamic onPush component should trigger change detection correctly

JiaLiPassion opened the pull request:
Close #36667, #39524, #39545.

`dynamic` and `bootstrap` components has a parent `host view` as the placeholder.
And also all the hooks are registered to the tView of the `host view`, not the component view.
Running change detection from the `host view` has several issues.

1. componentRef.changeDetectorRef.detectChanges()/markForCheck() will not update the component. Since componentRef.changeDetectorRef is the RootViewRef of the hostView.

```
@component({
  selector: 'dynamic',
  template: `{{name}}`,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class Dynamic  {
  name = 'initial name';
}

@component({
  selector: 'my-app',
  template: `
    <b>Expected</b>: "initial name" changes to "changed name" after 2s<br>
    <b>Actual</b>:<br>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent  {
  constructor(private _vcRef: ViewContainerRef,
     private _cfr: ComponentFactoryResolver, private _injector: Injector) {}

  ngOnInit() {
    const cmptRef = this._cfr.resolveComponentFactory(Dynamic).create(this._injector);
    this._vcRef.insert(cmptRef.hostView);

    setTimeout(() => {
        cmptRef.instance.name = 'changed name';
        console.log('changing name');
       // cmptRef.changeDetectorRef is the RootViewRef of the hostView
        cmptRef.changeDetectorRef.detectChanges();
    }, 2000);
  }
}
```

![Screen Shot 2020-11-03 at 0 28 37](https://user-images.githubusercontent.com/1442575/97886646-305ce580-1d6c-11eb-948a-d8a699c58004.png)


2. The hooks (such as ngOnCheck) is executed, but the component is not updated. Since the hooks are registered to the tView of the hostView.

```
import {
  Component,
  NgModule,
  ViewContainerRef,
  ComponentFactoryResolver,
  Injector,
  ChangeDetectionStrategy,
  DoCheck
} from "@angular/core";
import { BrowserModule } from "@angular/platform-browser";

import { timer } from "rxjs";

@component({
  selector: "dynamic",
  template: `
    {{ name }}
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class Dynamic implements DoCheck {
  name = "initial name";

  ngDoCheck() {
    console.log("docheck dynamic");
  }
}

@component({
  selector: "my-app",
  template: `
    <b>Expected</b>: "initial name" changes to "changed name" after 2s<br />
    <b>Actual</b>:<br />
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent {
  constructor(
    private _vcRef: ViewContainerRef,
    private _cfr: ComponentFactoryResolver,
    private _injector: Injector
  ) {}

  ngOnInit() {
    const cmptRef = this._cfr
      .resolveComponentFactory(Dynamic)
      .create(this._injector);
    this._vcRef.insert(cmptRef.hostView);

    setTimeout(() => {
      cmptRef.instance.name = "changed name";
      console.log("changing name");
      cmptRef.changeDetectorRef.detectChanges();
    }, 2000);
  }
}

@NgModule({
  imports: [BrowserModule],
  declarations: [AppComponent, Dynamic],
  entryComponents: [Dynamic],
  bootstrap: [AppComponent]
})
export class AppModule {}

```

![Screen Shot 2020-11-03 at 0 31 22](https://user-images.githubusercontent.com/1442575/97886671-3c48a780-1d6c-11eb-9a16-7364e81c02a2.png)

3. dev mode will only trigger change detection once for the `onPush` component.

```
      @Component({
        template: `
          <button (click)="noop()">Trigger change detection</button>
          <div>{{increment('componentView')}}</div>
        `,
        changeDetection: ChangeDetectionStrategy.OnPush
      })
      class App {
        increment() {
          counters++;
        }
        noop() {}
      }
      const fixture = TestBed.createComponent(App);
      fixture.detectChanges();
      expect(counters).toEqual(2); // current behavior is 1 
```
The reason is App is a OnPush component with a hostView, calling fixture.detectChanges() will run a detectChanges() and checkNoChanges() .
And the App component is onPush, so the checkNoChanges() run will ignore App component, so the counters will be 1
Which is not correct.
```
it('works', () => {
      @Component({
        template: '{{message}}',
        changeDetection: ChangeDetectionStrategy.OnPush
      })
      class OnPushCmp {
        @Input() message = 'initial';
        ngAfterViewInit() {
          this.message = 'updated';
        }
      }
      const comp = TestBed.configureTestingModule({declarations: [OnPushCmp]}).createComponent(OnPushCmp);
      expect(() => comp.detectChanges()).toThrow(); // current behavior is not throw.
    });
```
The reason is the same, the checkNoChanges() will skip the OnPushCmp, so the ExpressionChanged error is not thrown.

## This PR update several points.

1. When create a dynamic component, we create a host view and a component view.
We consider these 2 views a single unit,
if the component is OnPush, we set the Dirty flag to the host view to make it a an OnPush View,
 and set the CheckAlways flag to the component.

![Screen Shot 2020-11-03 at 20 50 03](https://user-images.githubusercontent.com/1442575/97982996-edf0e280-1e17-11eb-8b8b-191720b6c351.png)

2. About handling `OnPush` in `checkNoChanges()`, the current behavior is `OnPush` component will not execute the `checkNoChanges()` since the `Dirty` flag is cleared in the `detectChanges()`, so the `ExpressionChanged` error will not be thrown for `OnPush` component, which is a bug, but to fix it, it will be a `breaking` change for the existing apps, so this PR also add some logic to keep the current `buggy` behavior. so I also created another commit to add a `isOnPushCheckNoChangesEnabled` option to enable `checkNoChanges` for the `OnPush` component https://github.com/angular/angular/pull/39463/commits/4dcf5bdab2cfeded3adafba19329e7772ae217a7

JiaLiPassion committed:
fix(core): dynamic onPush component should trigger change detection correctly

Close #36667, #39524.

`dynamic` and `bootstrap` components has a parent `host view` as the placeholder.
And also all the hooks are registered to the tView of the `host view`, not the component view.
Running change detection from the `host view` has several issues.

1. componentRef.changeDetectorRef.detectChanges()/markForCheck() will not update the component.

```
@Component({
  selector: 'dynamic',
  template: `{{name}}`,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class Dynamic  {
  name = 'initial name';
}

@Component({
  selector: 'my-app',
  template: `
    <b>Expected</b>: "initial name" changes to "changed name" after 2s<br>
    <b>Actual</b>:<br>
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent  {
  constructor(private _vcRef: ViewContainerRef,
     private _cfr: ComponentFactoryResolver, private _injector: Injector) {}

  ngOnInit() {
    const cmptRef = this._cfr.resolveComponentFactory(Dynamic).create(this._injector);
    this._vcRef.insert(cmptRef.hostView);

    setTimeout(() => {
        cmptRef.instance.name = 'changed name';
        console.log('changing name');
       // cmptRef.changeDetectorRef is the RootViewRef of the hostView
        cmptRef.changeDetectorRef.detectChanges();
    }, 2000);
  }
}
```

2. The hooks (such as ngOnCheck) is executed, but the component is not updated.

```
import {
  Component,
  NgModule,
  ViewContainerRef,
  ComponentFactoryResolver,
  Injector,
  ChangeDetectionStrategy,
  DoCheck
} from "@angular/core";
import { BrowserModule } from "@angular/platform-browser";

import { timer } from "rxjs";

@Component({
  selector: "dynamic",
  template: `
    {{ name }}
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class Dynamic implements DoCheck {
  name = "initial name";

  ngDoCheck() {
    console.log("docheck dynamic");
  }
}

@Component({
  selector: "my-app",
  template: `
    <b>Expected</b>: "initial name" changes to "changed name" after 2s<br />
    <b>Actual</b>:<br />
  `,
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent {
  constructor(
    private _vcRef: ViewContainerRef,
    private _cfr: ComponentFactoryResolver,
    private _injector: Injector
  ) {}

  ngOnInit() {
    const cmptRef = this._cfr
      .resolveComponentFactory(Dynamic)
      .create(this._injector);
    this._vcRef.insert(cmptRef.hostView);

    setTimeout(() => {
      cmptRef.instance.name = "changed name";
      console.log("changing name");
      cmptRef.changeDetectorRef.detectChanges();
    }, 2000);
  }
}

@NgModule({
  imports: [BrowserModule],
  declarations: [AppComponent, Dynamic],
  entryComponents: [Dynamic],
  bootstrap: [AppComponent]
})
export class AppModule {}

```

This PR update several points.

1. When create a dynamic component, we create a host view and a component view.
We consider these 2 views a single unit,
if the component is OnPush, we set the Dirty flag to the host view to make it a an OnPush View,
 and set the CheckAlways flag to the component.

atscott commented:
At the very least, the current approach (by making the component view `CheckAlways`) is going to be a breaking change because it can produce new (valid) `ExpressionChangedAfterItHasBeenCheckedError`. `OnPush` components currently do not throw these errors when they should so by changing the component view to `CheckAlways`, the errors will now be produced. For reference, I had a PR a while back that would address this (#34443) but it is very break-y. This was also [proposed as a project](https://docs.google.com/document/d/1RJ1S4qdlwktm9Bx-3tfyIWE1tPZ-blAaSQrea601gYY/edit#heading=h.yg86x3m6xp64) at some point but must have been lost. I thought it was an item in the planning spreadsheet, but I was either mistaken or it got dropped/lost.

Example test (which passes without your changes but fails with your changes):
```

    it('works', () => {
      @Component({
        template: '{{message}}',
        changeDetection: ChangeDetectionStrategy.OnPush
      })
      class OnPushCmp {
        @Input() message = 'initial';

        ngAfterViewInit() {
          this.message = 'updated';
        }
      }
      const comp = TestBed.configureTestingModule({declarations: [OnPushCmp]}).createComponent(OnPushCmp);
      expect(() => comp.detectChanges()).not.toThrow();
    });
```

JiaLiPassion committed:
test(core): update dynamic onPush component change detection test

Update test cases, current onPush component test will consider comp.detectChanges() not update
the component. This commit update the related case to make comp.detectChanges() update the component.

JiaLiPassion committed:
fix(core): keep onPush not run checkNoChanges behavior

There is another issue related to OnPush component.

```
it('works', () => {
      @Component({
        template: '{{message}}',
        changeDetection: ChangeDetectionStrategy.OnPush
      })
      class OnPushCmp {
        @Input() message = 'initial';
        ngAfterViewInit() {
          this.message = 'updated';
        }
      }
      const comp = TestBed.configureTestingModule({declarations: [OnPushCmp]}).createComponent(OnPushCmp);
      expect(() => comp.detectChanges()).toThrow(); // current behavior is not throw.
    });
```

The current behavior is OnPush component will not really run `checkNoChanges()` since `detectChanges()`
already clear the `Dirty` flag, so no `ExpressionChanged` error will be thrown from the onPush component.

This PR also fix this issue, but this will become a very breaking change, and cause a lot of apps throw the
`ExpressionChanged` errors. So this commit add the logic to ignore the `checkNoChanges()` code path for the
onPush component.

JiaLiPassion committed:
test(core): update golden symbols

Update global symbols for test to add the `isLRootViewAndNeedsCheck`
function

JiaLiPassion commented:
@atscott, I added a commit to keep the current behavior, so the current behavior is `onPush` component doesn't run `checkNoChanges()` at all, since the `Dirty` flag is cleared after `detectChanges()`, so this commit https://github.com/angular/angular/pull/39463/commits/91e1521f8a3db954539958b49befa35efb415fd8 ignore the `checkNoChanges()` code path if the view is onPush.

atscott reviewed (changes requested):
@JiaLiPassion I'm a little confused by some of the changes in 91e1521 . Is it necessary to add this special case check to so many places? I would have thought that it would only have to affect the root view, but there are changes to `refreshComponent` and `checkNoChangesInternal`.  It would be good to know what tests were failing before - I was trying `yarn bazel test //packages/core/test/acceptance --config=ivy` and all of those tests were still passing if I reverted some of the changes.

In addition, the return value in `isLRootViewAndNoNeedToCheck` for the special case seems to be backwards: since your intent was to skip `OnPush` components when in `checkNoChanges`, then the logical return value should be `true` to indicate that there is "no need to check".  I think it would be better to change the entire function to be `isLRootViewAndNeedsCheck` rather than dealing with how to interpret a double negative.

Lastly, I'm not really convinced that this is the correct solution. The rules for checking a view are correct and logical at the moment - this commit introduces a special path to work around a bug in the current system instead of fixing the root problem that came up.

I think it would also be good to get @mhevery's thoughts on this as well and/or add this to the agenda for our sync.

JiaLiPassion commented:
@atscott. 

> I'm a little confused by some of the changes in 91e1521 . Is it necessary to add this special case check to so many places? I would have thought that it would only have to affect the root view, but there are changes to refreshComponent and checkNoChangesInternal.

Yeah, you are right, `refreshComponent` doesn't need to add this logic, I am not sure about `checkNoChangesInternal`, I will check it again.

> In addition, the return value in isLRootViewAndNoNeedToCheck for the special case seems to be backwards: since your intent was to skip OnPush components when in checkNoChanges, then the logical return value should be true to indicate that there is "no need to check". I think it would be better to change the entire function to be isLRootViewAndNeedsCheck rather than dealing with how to interpret a double negative.

Yes, you are right, I updated this part and also added test cases.

> Lastly, I'm not really convinced that this is the correct solution. The rules for checking a view are correct and logical at the moment - this commit introduces a special path to work around a bug in the current system instead of fixing the root problem that came up.

Yeah, I totally understand, so if we want to fix this issue, just like we discussed, I think remove this `special` logic to handle `onPush checkNoChanges()` is the correct behavior, but it will break so many existing applications. I believe this `special handling` make the current behavior the same to give us some time. 
Also I think maybe we can provide a `flag` such like `enableOnPushNoChangeCheck` when bootstrap module to let user use the correct behavior.

JiaLiPassion committed:
feat(core): add isOnPushCheckNoChangesEnabled option

Currently when running `detectChanges()` for OnPush component in dev mode,
the `checkNoChanges()` will not be executed since the `Dirty` flag is cleared
after the `detectChanges()` call. This issue can be fixed but becomes a breaking changes
and impact a lot of applications which will throw `ExpressionChanged` error after
the fix.

@Component({
   template: '{{message}}',
   changeDetection: ChangeDetectionStrategy.OnPush
})
class OnPushCmp {
  @Input() message = 'initial';
  ngAfterViewInit() {
    this.message = 'updated';
  }
}
const comp = TestBed.configureTestingModule({declarations:
  OnPushCmp]}).createComponent(OnPushCmp);

Consider this case, the OnPushCmp should throw `ExpressionChanged` error since
it update the data in `ngAfterViewInit` hook, but it didn't throw since `OnPush`
component didn't run the `checkNoChange` code path. The behavior is not correct, but
fix it will cause a big breaking change.

So now the behavior is kept the same with before to avoid the breaking changes, and
here we have a new flag `isOnPushCheckNoChangesEnabled` to enable the fix, so the
`OnPush` component can enter the `checkNoChanges()` code path to verify there are unintended
changes or not.

By default the value is `false`.

JiaLiPassion commented:
I created another commit to add a `isOnPushCheckNoChangesEnabled` option to enable `checkNoChanges` for the `OnPush` component https://github.com/angular/angular/pull/39463/commits/4dcf5bdab2cfeded3adafba19329e7772ae217a7

atscott commented on packages/core/src/application_ref.ts:
```diff
@@ -266,6 +267,22 @@ export interface BootstrapOptions {
+   *
+   * By default the value is `false`.
+   */
+  isOnPushCheckNoChangesEnabled?: boolean;
```
So I think the problem with this flag is that it's really misleading. Enabling this doesn't `'fix'` the issue with `OnPush` components in most cases. It only changes the behavior when the `OnPush` component is the root component. However, it does nothing to fix `checkNoChanges` mode for non-root `OnPush` components, which will _still_ not be checked because their dirty flags have been cleared.

Having this flag is one of the possible solutions to add to a _complete_ fix of the `OnPush` components in `checkNoChanges` mode, but I don't think it should be added for this partial fix because:
* It's misleading, as mentioned above
* It doesn't avoid the extra "special case" logic as discussed in a previous comment on this PR. It actually just adds another condition to the special case logic, which is potentially even more confusing.

Another possible way to avoid the new `ExpressionChanged` errors would be to make them "warnings" if the component is `OnPush` rather than errors.

So I think again, the way forward is:
* First fix the behavior for `OnPush` components in `checkNoChanges` mode in all cases. The options here are numerous for how to approach the `ExpressionChanged` error:
  * Keep it an error and make the change in  a major release. Patch g3 to not produce the error
  * Make it a warning instead
  * Add a flag like you have here to control the behavior
* Ignore the underlying issue and add the special case logic. I think if this option is chosen, we _should not_ add this flag because it just adds to the confusion and special case logic.

Either way, I'll run another presubmit to see if there are other types of failures caused by this change. 

Sorry to keep pushing back on this, but I want to be sure we get it right. In my opinion, this is a pretty significant change. It'll be good to see how others feel about this when we discuss in the team sync.

atscott replied:
[presubmit](https://test.corp.google.com/ui#id=OCL:340242567:BASE:340666852:1604511736133:24f3550) is showing several failures and require additional investigation.

JiaLiPassion replied:
@atscott , thank you for the review, and you are right, this flag not fixing the whole issue, it only fixed the `dynamic onpush checkNoChanges` issue for now, I will update the commit to try to fix the `normal onpush checkNoChanges` issue too. 

> 
> First fix the behavior for OnPush components in checkNoChanges mode in all cases. The options here are numerous for how to approach the ExpressionChanged error:
> Keep it an error and make the change in a major release. Patch g3 to not produce the error
> Make it a warning instead
> Add a flag like you have here to control the behavior
> Ignore the underlying issue and add the special case logic. I think if this option is chosen, we should not add this flag because it just adds to the confusion and special case logic.

Got it, I think this is reasonable, I will continue to work on it!

> Sorry to keep pushing back on this, but I want to be sure we get it right. In my opinion, this is a pretty significant change. It'll be good to see how others feel about this when we discuss in the team sync.

Yeah, no problem, thank you again for the detailed review, I have added an item in the team sync meeting note.
